File: Java/D41/d41p3.java
/*
 * Given an array of integer elements, ele[], YOur task is to find and print 
the number of the subarrays, which gives you an even value after multiplication
of the elements in it.

Input Format:
-------------
Line-1: An integer N, number of elements.
Line-2: Space separated elements.

Output Format:
--------------
Print number of possible subarrays.


Sample Input-1:
---------------
4
5 8 9 17

Sample Output-1:
----------------
6

Explanation:
------------
5 x 8 = 40
5 x 8 x 9 = 360
5 x 8 x 9 x 17 = 6120
8 = 8
8 x 9 = 72
8 x 9 x 17 = 1224
Hence we got 6 even subarray multiplications.



Sample Input-2:
---------------
3
5 7 11

Sample Output-2:
----------------
0

Explanation:
------------
No subarrays which will give you even value after multiplication.

*/

import java.util.*;

/**
 * d41p3
 */
public class d41p3 {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int arr[] = new int[n];
        boolean flag = true;
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
            flag = flag && arr[i] % 2 == 0;
        }
        sc.close();
        if (flag) {
            System.out.println(0);
            return;
        }
        int count = 0;
        for (int i = 0; i < n; i++) {
            int mul = 1;
            for (int j = i; j < n; j++) {
                mul *= arr[j];
                if (mul % 2 == 0) {
                    count++;
                }
            }
        }
        System.out.println(count);
    }
}

File: CPP/D41/d41p3.cpp
/*

In a forest, There are N redwoord trees in a row.
You are given the heights of the trees as heights[],

You are task is to find the longest tree arrangement as follows:
	- Minimum size of the tree arrangement is 3.
	- And there exist a Tree-'i' with heights[i], where 0 < i < N-1.
		- heights[0] < heights[1] < heights[2] <...< heights[i] and
		-  heights[i] > heights[i+1] > heights[i+2] >...>heights[N-1] 

And return the length of the longest tree arrangement.
If there is no such arrangement, return 0.

Input Format:
-------------
Line-1: An integer N, number of elements.
Line-2: N space separated integers, value of the elements.

Output Format:
--------------
Print an integer result.


Sample Input-1:
---------------
8
4 2 5 7 4 2 3 6

Sample Output-1:
----------------
5

Explanation:
------------
The longest tree arrangement is : 2 5 7 4 2


Sample Input-2:
---------------
4
2 4 5 7

Sample Output-2:
----------------
0




*/

#include<bits/stdc++.h>
using namespace std;

int main(){
    int n;
    cin>>n;
    vector<int> arr(n,0);
    for(int i=0;i<n;++i){
        cin>>arr[i];
    }
    vector<int> lis(n,0);
    for(int i=1;i<n;++i){
            if(arr[i]>arr[i-1]){
                lis[i]=lis[i-1]+1;
        }
    }
    vector<int> lds(n,0);
    for(int i=n-2;i>=0;--i){
            if(arr[i]>arr[i+1]){
                lds[i]=lds[i+1]+1;
        }
    }
    int maxi=0;
    for(int i=0;i<n;++i){
        if(lis[i]>0 && lds[i]>0){
            maxi=max(maxi,lis[i]+lds[i]+1);
        }
    }
    cout<<maxi;
    return 0;
}